src/mapping/query/function/sha3.rs

```diff
@@ -0,0 +1,140 @@
+use super::prelude::*;
+use digest::DynDigest;
+use sha3::{Digest, Sha3_224, Sha3_256, Sha3_384, Sha3_512};
+
+#[derive(Debug)]
+pub(in crate::mapping) struct Sha3Fn {
+    query: Box<dyn Function>,
+    variant: Option<Box<dyn Function>>,
+}
+
+impl Sha3Fn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, variant: Option<&str>) -> Self {
+        let variant = variant.map(|v| Box::new(Literal::from(Value::from(v))) as _);
+
+        Self { query, variant }
+    }
+}
+
+impl Function for Sha3Fn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+
+        let variant = optional!(ctx, self.variant, Value::Bytes(v) => v)
+            .map(|v| String::from_utf8_lossy(&v).into_owned());
```

Comment:
Since we know the exact variants we want to match this parameter against, I'm wondering if there is a way we can avoid the whole `from_utf8` call altogether. 

You could store a lazy_static `SHA3-224 = Byte::from("SHA3-223")` and then just match `variant` against that.

Overhead will be tiny, so its not a big deal, but also I don't think it has any cost so may be worth it.

Reply:
Yeah, there are definitely some things we can change here.

I'm going to leave this as is, as I want to get to the finish line of implementing all requested functions first. There is also almost no gain in removing an extra allocation, compared to the hashing itself that happens in these functions.

Still, I'm keeping a list of things we can improve generally in all functions, so if there's still some time left once they are all in, I'm going to circle back and make some minor tweaks in all of them at once.

Reply:
Went ahead and did the same thing I did for the sha2 function, which removes the allocations.